draft-irtf-cfrg-vdaf.md

```diff
@@ -2384,140 +2433,124 @@ def joint_rand(Prio3, k_joint_rand_seed):
-        return (meas_share, proof_share, None)
-    k_blind, encoded = encoded[:l], encoded[l:]
-    if len(encoded) != 0:
-        raise ERR_DECODE
-    return (meas_share, proof_share, k_blind)
 ~~~
+struct {
+    Prio3Seed k_joint_rand_parts[Prio3.SHARES];
+} Prio3PublicShareWithJointRand;
+~~~
+
+#### Input share
 
-Next, the methods below are used for encoding and decoding the helpers' (i.e.,
-non-leader) VDAF input shares. Each consists of PRG seeds that are expanded
-into the measurement and proof shares.
+Just as for the public share, the encoding of the input shares dpends on
+whether joint randomness is used. If so, then each input share includes the
+Aggregator's blind for generating its joint randomness part.
+
+In addition, the encoding of the input shares depends on which aggregator is
+receiving the message. If the aggregator ID is `0`, then the input share
+includes the full measurement and proof share. Otherwise, if the aggregator ID
+is greater than `0`, then the measurement and proof shares are represented by
+PRG seeds. We shall call the former the "Leader" and the latter the "Helpers".
+
+In total there are four variants of the input share. When joint randomness is
+not used, the Leader's share is structured as follows:
 
 ~~~
-def encode_helper_share(Prio3,
-                        k_meas_share,
-                        k_proof_share,
-                        k_blind):
-    encoded = Bytes()
-    encoded += k_meas_share
-    encoded += k_proof_share
-    if Prio3.Flp.JOINT_RAND_LEN > 0:
-        encoded += k_blind
-    return encoded
+struct {
+    Prio3.Flp.Field meas_share[Prio3.Flp.MEAS_LEN];
+    Prio3.Flp.Field proof_share[Prio3.Flp.PROOF_LEN];
+} Prio3LeaderShare;
+~~~
+
+Finally, when joint randomness is not used, the Helpers' shres are structured
+as follows:
 
-def decode_helper_share(Prio3, agg_id, encoded):
-    l = Prio3.Prg.SEED_SIZE
-    k_meas_share, encoded = encoded[:l], encoded[l:]
-    meas_share = Prio3.helper_meas_share(agg_id, k_meas_share)
-    k_proof_share, encoded = encoded[:l], encoded[l:]
-    proof_share = Prio3.helper_proof_share(agg_id, k_proof_share)
-    if Prio3.Flp.JOINT_RAND_LEN == 0:
-        if len(encoded) != 0:
-            raise ERR_DECODE
-        return (meas_share, proof_share, None)
-    k_blind, encoded = encoded[:l], encoded[l:]
-    if len(encoded) != 0:
-        raise ERR_DECODE
-    return (meas_share, proof_share, k_blind)
+~~~
+struct {
+    Prio3Seed k_meas_share;
+    Prio3Seed k_proof_share;
+} Prio3HelperShare;
 ~~~
 
-Next, the methods below are used for encoding and decoding the VDAF public share.
+When joint randomness is used, the Leader's input share is structured as
+follows:
 
 ~~~
-def encode_public_share(Prio3,
-                        k_joint_rand_parts):
-    encoded = Bytes()
-    if Prio3.Flp.JOINT_RAND_LEN > 0:
-        encoded += concat(k_joint_rand_parts)
-    return encoded
+struct {
+    Prio3.Flp.Field meas_share[Prio3.Flp.MEAS_LEN];
```

Comment:
That's a good idea. One annoyance is that we'd have to name the field, so instead of 
`leader_share.meas_share` we'd refer to `leader_share.inner.meas_share`:

```
struct {
    Prio3LeaderShare inner;
    Prio3Seed k_blind;
} Prio3LeaderShareWithJointRand;
```

But I think the dedup is really nice so I took your suggestion.